source/adapters/level_zero/v2/latency_tracker.hpp

```diff
@@ -0,0 +1,104 @@
+//===--------- ur_latency_tracker.cpp - common ---------------------------===//
+//
+// Copyright (C) 2024 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <limits>
+#include <memory>
+
+#include "logger/ur_logger.hpp"
+
+namespace v2 {
+
+static inline bool trackLatency = []() {
+  return std::getenv("UR_ENABLE_LATENCY_TRACKING") != nullptr;
+}();
+
+class rolling_stats {
+public:
+  rolling_stats(const char *name) : name(name) {}
+
+  ~rolling_stats() {
+    if (trackLatency) {
+      logger::info("[{}] average latency: {}ns", name, estimate());
+      logger::info("[{}] number of samples: {}", name, count());
+    }
+  }
+
+  // track latency by taking the value of duration directly.
+  void trackValue(double value) {
+    // This is a highly unlikely scenario where
+    // cnt_ reaches numerical limits. Skip update
+    // of the rolling average anymore.
+    if (cnt_ == std::numeric_limits<uint64_t>::max()) {
+      cnt_ = 0;
+      return;
+    }
+    auto ratio = static_cast<double>(cnt_) / (cnt_ + 1);
+    avg_ *= ratio;
+    ++cnt_;
+    avg_ += value / cnt_;
+  }
+
+  // Return the rolling average.
+  double estimate() { return avg_; }
+
+  // Number of samples tracked.
+  uint64_t count() { return cnt_; }
+
+private:
+  const char *name;
+  double avg_{0};
```

Comment:
why is this a double and not just an integer value?

Reply:
I would still need to cast it to double for calculations in trackValue I think. I'm not sure if it matters really. But I have changed the estimate() return value to be uint64_t

Reply:
> I would still need to cast it to double for calculations in trackValue I think. I'm not sure if it matters really.

Well, I just prefer not using floating point math if we don't have to, especially for something so simple as `sum += value; ++cnt; avg = sum / cnt;`. But yeah, I don't think matters all that much.